ls1intum/Artemis
# `Lectures`: Allow instructors to automatically split lecture file units

### Checklist
#### General


#### Server
#### Client

### Motivation and Context
Closes #4966

Instructors have to create multiple file units after they create the lecture. This process is time-consuming and cumbersome for instructors. Therefore an automatic way of unit creation is introduced in this PR. 

### Description


**Server**
Two new endpoints are introduced in this PR:
- POST lectures/:lectureId/process-units : It prepares attachment units information that need to be displayed into the table
- POST lectures/:lectureId/attachment-units/split : Creates new attachment units based on provided information

**LectureUnitProcessingService**:  [Apache PDFBox® - A Java PDF Library](https://pdfbox.apache.org/) is used to split the file into units.
**AttachmentUnitService**:  Added new function that creates new attachment units for the given lecture.

**Client**

- Existing component (lecture-update.component) is modified where it gives the instructor an option to upload the lecture file and proceed into automatic splitting.
- New component (attachment-units.component) is added where instructors will be able to see unit information in a table and then submit to create the attachment units.

 


### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Course with lectures

1. Log in to Artemis
2. Navigate to course management
3. Go to lectures page of the course
4. Click Create Lecture
5. Check the checkbox (Automatic unit processing)
6. You should see file upload option and Process units button on the right
7. After you upload the file click Process units
8. Then you will get redirected to a new page with a table with information filled
9. Check if the table contains correct Unit Name, Start Page and End Page. 
10. Then click submit, you will get redirected to Lecture Units page where you should be able to see newly created lecture units

This a lecture file where you can test this PR
[ITP2223 L05 Object Orientation I.pdf](https://github.com/ls1intum/Artemis/files/10396515/ITP2223.L05.Object.Orientation.I.pdf)


### Review Progress






#### Code Review
#### Manual Tests

### Test Coverage




**Server**
| Class/File | Branch | Line |
|------------|-------:|-----:|
| AttachmentUnitService.java | 70% | 97% |
| LectureUnitProcessing.java | 98% | 91% |
| AttachmentUnitResource.java | 50 | 75% |
| LectureUnitInformationDTO.java | n/a | 100% |
| LectureUnitDTO.java | n/a | 100% |
| LectureUnitSplitDTO.java | n/a | 100% |

**Client**
| Class/File | Branch | Line |
|------------|-------:|-----:|
| lecture-update.component.ts | 82.97%| 92.59% |
| attachment-units.component.ts | 51.66%| 88.37% |
| attachmentUnit.service.ts | 70%| 100% |


### Screenshots



Now, Instructors will have an option to choose if they want to automatically split the lecture.
![lecture-update](https://user-images.githubusercontent.com/26312757/212767749-c2a958c5-98ba-40fb-9e5e-b2901625e794.png)

After they click Process units, they will get redirected to a new page where they will see the table filled with information.
![attachUnits](https://user-images.githubusercontent.com/26312757/212767647-eb2d5ced-fdfb-4529-a925-f46d58c1fbe6.png)


If the file you are about to split, does not contain Outline slides, no unit will be detected so it should look like the following:
![noUnitsDetected](https://user-images.githubusercontent.com/26312757/212767704-ba949fc1-8db5-4400-92a2-d5e5866dff53.png)
